### PR TITLE
SQSERVICES-1167-backend-2nd-factor-for-add-new-client

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/SQSERVICES-1167

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] **changelog.d** contains the following bits of information ([details](https://github.com/wireapp/wire-server/blob/develop/docs/developer/changelog.md)):
   - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

File: services/brig/test/integration/API/User/Client.hs [hunk 1]
```diff
@@ -96,6 +110,67 @@ tests _cl _at opts p b c g =
...
+    const (Just c) === responseJsonMaybe
+
+testAddGetClientMissingCode :: Brig -> Galley -> Http ()
+testAddGetClientMissingCode brig galley = do
+  (u, tid) <- Util.createUserWithTeam' brig
+  let uid = userId u
+  let addClient' codeValue = addClient brig uid (defNewClientWithVerificationCode codeValue PermanentClientType [head somePrekeys] (head someLastPrekeys))
+
+  Util.setTeamSndFactorPasswordChallenge galley tid Public.TeamFeatureEnabled
+  addClient' Nothing !!! do
```
Comment (fisx) [inline services/brig/test/integration/API/User/Client.hs:139]:
```suggestion
  Util.generateVerificationCode brig (Public.SendVerificationCode Public.Login email)
  addClient' Nothing !!! do
```

File: services/brig/test/integration/API/User/Client.hs [hunk 2]
```diff
@@ -96,6 +110,67 @@ tests _cl _at opts p b c g =
...
+    const (Just "code-authentication-required") === fmap Error.label . responseJsonMaybe
+
+testAddGetClientWrongCode :: Brig -> Galley -> Http ()
+testAddGetClientWrongCode brig galley = do
+  (u, tid) <- Util.createUserWithTeam' brig
+  let uid = userId u
+  let addClient' codeValue = addClient brig uid (defNewClientWithVerificationCode codeValue PermanentClientType [head somePrekeys] (head someLastPrekeys))
+
+  Util.setTeamSndFactorPasswordChallenge galley tid Public.TeamFeatureEnabled
+  let wrongCode = Code.Value $ unsafeRange (fromRight undefined (validate "123456"))
```
Comment (fisx) [inline services/brig/test/integration/API/User/Client.hs:150]:
```suggestion
  Util.generateVerificationCode brig (Public.SendVerificationCode Public.Login email)
  let wrongCode = Code.Value $ unsafeRange (fromRight undefined (validate "123456"))
```

File: services/brig/test/integration/API/User/Client.hs [hunk 3]
```diff
@@ -96,6 +110,67 @@ tests _cl _at opts p b c g =
...
+  (u, tid) <- Util.createUserWithTeam' brig
+  let uid = userId u
+  let Just email = userEmail u
+  let checkLoginSucceeds b = login brig b PersistentCookie !!! const 200 === statusCode
+  let addClient' codeValue = addClient brig uid (defNewClientWithVerificationCode codeValue PermanentClientType [head somePrekeys] (head someLastPrekeys))
+
+  Util.setTeamSndFactorPasswordChallenge galley tid Public.TeamFeatureEnabled
+  Util.generateVerificationCode brig (Public.SendVerificationCode Public.Login email)
+  k <- Code.mkKey (Code.ForEmail email)
+  (fmap Code.codeValue -> codeValue) <- lookupCode db k Code.AccountLogin
```
Comment (fisx) [inline services/brig/test/integration/API/User/Client.hs:166]: see above.

File: services/brig/test/integration/API/User/Client.hs [hunk 4]
```diff
@@ -96,6 +110,67 @@ tests _cl _at opts p b c g =
...
+  checkLoginSucceeds $ PasswordLogin (LoginByEmail email) defPassword (Just defCookieLabel) codeValue
+  -- wait > 5 sec for the code to expire (assumption: setVerificationTimeout in brig.integration.yaml is set to <= 5 sec)
+  threadDelay $ (5 * 1000000) + 1000
```
Comment (fisx) [inline services/brig/test/integration/API/User/Client.hs:169]: see above.

File: services/brig/src/Brig/Options.hs [hunk 5]
```diff
@@ -582,7 +582,7 @@ setDefaultUserLocale :: Settings -> Locale
 setDefaultUserLocale = fromMaybe defaultUserLocale . setDefaultUserLocaleInternal
 
 defVerificationTimeout :: Code.Timeout
-defVerificationTimeout = Code.Timeout (60 * 10) -- 10 minutes
+defVerificationTimeout = Code.Timeout (60 * 15) -- 15 minutes
```
[the remark this answers is not in the document]
Comment (battermann) [reply services/brig/src/Brig/Options.hs:585]: It's a requirement in the ticket SQSERVICES-1167.

[the remark this answers is not in the document]
Comment (fisx) [reply services/brig/src/Brig/Options.hs:585]: Good point, I'll change it both there and here.

**Comment by fisx** [comment]
all tests pass locally.

**Review by fisx — APPROVED** [review]

**Pull request merged**